File: services/tickets/src/handlers/ticketadmin.rs
use crate::repository::{
    guild::{GuildRepository, GuildService, UpdateGuildData},
    ticket::{TicketRepository, TicketService},
};
use async_trait::async_trait;
use duvua_framework::{
    builder::{button_action_row::CreateActionRow, interaction_response::InteractionResponse},
    errors::BotError,
    handler::{CommandHandler, CommandHandlerData, CustomCommandType},
    utils::{get_option, get_sub_command, send_message},
};
use serenity::{
    builder::{
        CreateApplicationCommand, CreateApplicationCommandOption, CreateButton, CreateEmbed,
        CreateMessage,
    },
    http::Http,
    model::prelude::{
        application_command::{ApplicationCommandInteraction, CommandDataOption},
        command::CommandOptionType,
        component::ButtonStyle,
        ReactionType,
    },
    prelude::Context,
};
use std::sync::Arc;

pub struct TicketAdminCommandHandler {
    ticket_repo: Arc<TicketService>,
    guild_repo: Arc<GuildService>,
    data: &'static CommandHandlerData,
}

impl TicketAdminCommandHandler {
    pub fn new(ticket_repo: Arc<TicketService>, guild_repo: Arc<GuildService>) -> Self {
        Self {
            guild_repo,
            ticket_repo,
            data: Box::leak(Box::new(build_data())),
        }
    }

    pub async fn handle_toggle_guild_tickets(
        &self,
        guild_id: u64,
        enabled: bool,
    ) -> Result<InteractionResponse, BotError> {
        self.guild_repo
            .update(
                guild_id,
                UpdateGuildData {
                    allow_multiple: None,
                    channel_category: None,
                    enable_tickets: Some(enabled),
                },
            )
            .await?;

        let msg = if enabled {
            "Tickets habilidatos no servidor"
        } else {
            "Ticket desabilidatos no servidor"
        };

        Ok(InteractionResponse::with_content(msg))
    }

    pub async fn handle_delete_ticket_by_id(
        &self,
        http: impl AsRef<Http>,
        id: String,
    ) -> Result<InteractionResponse, BotError> {
        let ticket = self.ticket_repo.get(id.clone()).await?;

        match http.as_ref().delete_channel(ticket.channel_id as u64).await {
            Ok(c) => log::info!("Channel {} deleted", c.id().0),
            Err(e) => log::warn!("Failed to delete channel: {e}"),
        }

        self.ticket_repo.delete(id).await?;

        Ok(InteractionResponse::with_content(format!(
            "Ticket de <@{}> deletado com sucesso",
            ticket.user_id
        )))
    }

    pub async fn handle_delete_ticket_by_options(
        &self,
        http: impl AsRef<Http>,
        options: &Vec<CommandDataOption>,
    ) -> Result<InteractionResponse, BotError> {
        let id = get_option(options, "id")
            .ok_or(BotError::OptionNotProvided("id"))?
            .value
            .ok_or(BotError::InvalidOption("id"))?
            .as_str()
            .ok_or(BotError::InvalidOption("id"))?
            .to_owned();

        self.handle_delete_ticket_by_id(http, id).await
    }

    pub async fn handle_add_permanent(
        &self,
        http: impl AsRef<Http>,
        user_id: u64,
        message: Option<String>,
        channel_id: u64,
    ) -> Result<InteractionResponse, BotError> {
        let message = match message {
            Some(message) => format!("{message}\n- <@{user_id}>"),
            None => "Clique no botão para criar um ticket!".to_owned(),
        };

        let msg = CreateMessage::default()
            .set_embed(
                CreateEmbed::default()
                    .title("Tickets")
                    .description(message)
                    .to_owned(),
            )
            .set_components(
                CreateActionRow::default()
                    .add_button(
                        CreateButton::default()
                            .style(ButtonStyle::Primary)
                            .label("Criar Ticket")
                            .emoji(ReactionType::Unicode("🎫".to_owned()))
                            .custom_id("ticket")
                            .to_owned(),
                    )
                    .to_components(),
            )
            .to_owned();

        send_message(http, msg, channel_id)
            .await
            .or_else(|_| Err(BotError::FailedToSendChannelMessage))?;

        Ok(InteractionResponse::with_content("Mensagem enviada"))
    }

    pub async fn handle_add_permanent_by_options(
        &self,
        http: impl AsRef<Http>,
        user_id: u64,
        options: &Vec<CommandDataOption>,
    ) -> Result<InteractionResponse, BotError> {
        let channel = get_option(options, "channel")
            .ok_or(BotError::OptionNotProvided("channel"))?
            .value
            .ok_or(BotError::InvalidOption("channel"))?;

        let channel = channel
            .as_str()
            .ok_or(BotError::InvalidOption("channel"))?
            .parse()
            .or_else(|_| Err(BotError::InvalidOption("channel")))?;

        let message = match get_option(options, "message") {
            Some(message) => {
                let value = message.value.ok_or(BotError::InvalidOption("message"))?;
                let value = value.as_str().ok_or(BotError::InvalidOption("message"))?;

                Some(value.to_owned())
            }
            None => None,
        };

        self.handle_add_permanent(http, user_id, message, channel)
            .await
    }

    pub async fn handle_delete_all(
        &self,
        http: &Arc<Http>,
        guild_id: u64,
        user_id: u64,
    ) -> Result<InteractionResponse, BotError> {
        let tickets = self
            .ticket_repo
            .get_by_member(guild_id, user_id, 11)
            .await?;

        if tickets.len() > 10 {
            return Ok(InteractionResponse::with_content(
                "O usuário tem mais de 10 tickets, exclua eles manualmente",
            ));
        }

        let deleted_count = self.ticket_repo.delete_by_member(guild_id, user_id).await?;

        for ticket in tickets {
            let http = http.clone();

            tokio::spawn(async move {
                match http.delete_channel(ticket.channel_id as u64).await {
                    Ok(c) => log::info!("Channel {} on guild {guild_id} deleted", c.id().0),
                    Err(e) => log::warn!("Failed to delete channel: {e}"),
                }
            });
        }

        Ok(InteractionResponse::with_content(format!(
            "{deleted_count} tickets excluídos com sucesso"
        )))
    }
}

#[async_trait]
impl CommandHandler for TicketAdminCommandHandler {
    async fn handle_command(
        &self,
        ctx: &Context,
        interaction: &ApplicationCommandInteraction,
    ) -> Result<(), BotError> {
        let member_permissions = interaction
            .member
            .as_ref()
            .ok_or(BotError::CommandIssuedOutOfGuild)?
            .permissions
            .ok_or(BotError::CommandPermissionDenied)?;

        let guild_id = interaction
            .guild_id
            .ok_or(BotError::CommandIssuedOutOfGuild)?
            .0;

        if !member_permissions.administrator() {
            return Err(BotError::CommandPermissionDenied);
        }

        let sub_command =
            get_sub_command(&interaction.data.options).ok_or(BotError::SomethingWentWrong)?;

        let res = match sub_command.name.as_str() {
            "disable" => self.handle_toggle_guild_tickets(guild_id, false).await?,
            "enable" => self.handle_toggle_guild_tickets(guild_id, true).await?,
            "delete" => {
                self.handle_delete_ticket_by_options(&ctx.http, &sub_command.options)
                    .await?
            }
            "delete-all" => {
                self.handle_delete_all(&ctx.http, guild_id, interaction.user.id.0)
                    .await?
            }
            "add-permanent" => {
                self.handle_add_permanent_by_options(
                    &ctx.http,
                    interaction.user.id.0,
                    &sub_command.options,
                )
                .await?
            }
            _ => return Err(BotError::InvalidOption("sub-command")),
        };

        res.respond(&ctx.http, interaction.id.0, &interaction.token)
            .await
    }

    fn get_data(&self) -> &'static CommandHandlerData {
        self.data
    }
}

#[inline]
fn build_data() -> CommandHandlerData {
    CommandHandlerData {
        command_data: Some(build_data_command()),
        custom_id: None,
        needs_defer: false,
        kind: CustomCommandType::Config,
    }
}

fn build_data_command() -> CreateApplicationCommand {
    CreateApplicationCommand::default()
        .name("ticketadmin")
        .description("Comandos administrativos de tickets")
        .description_localized("en-US", "Ticket administrative commands")
        .add_option(
            CreateApplicationCommandOption::default()
                .kind(CommandOptionType::SubCommand)
                .name("disable")
                .description("Desabilita a funcionalidade de tickets no servidor")
                .description_localized("en-US", "Disables the ticket functionality on the server")
                .to_owned(),
        )
        .add_option(
            CreateApplicationCommandOption::default()
                .kind(CommandOptionType::SubCommand)
                .name("enable")
                .description("Habilita a funcionalidade de tickets no servidor")
                .description_localized("en-US", "Enables the ticket functionality on the server")
                .to_owned(),
        )
        .add_option(
            CreateApplicationCommandOption::default()
                .kind(CommandOptionType::SubCommand)
                .name("add-permanent")
                .description("Posta uma mensagem com um botão para criar tickets no servidor")
                .description_localized(
                    "en-US",
                    "Posts a message with a button to create tickets in the server",
                )
                .add_sub_option(
                    CreateApplicationCommandOption::default()
                        .kind(CommandOptionType::Channel)
                        .name("channel")
                        .description("O canal para enviar a mensagem")
                        .description_localized("en-US", "The channel the message will be sent")
                        .required(true)
                        .to_owned(),
                )
                .add_sub_option(
                    CreateApplicationCommandOption::default()
                        .kind(CommandOptionType::String)
                        .name("message")
                        .description("A mensagem que será enviada")
                        .description_localized("en-US", "The message")
                        .to_owned(),
                )
                .to_owned(),
        )
        .add_option(
            CreateApplicationCommandOption::default()
                .kind(CommandOptionType::SubCommand)
                .name("delete")
                .description("Deleta o ticket de um membro por id")
                .description_localized("en-US", "Deletes one member's ticket by id")
                .add_sub_option(
                    CreateApplicationCommandOption::default()
                        .kind(CommandOptionType::String)
                        .name("id")
                        .description("O id do ticket que deseja deletar")
                        .description_localized("en-US", "The id of the ticket you want to delete")
                        .required(true)
                        .to_owned(),
                )
                .to_owned(),
        )
        .add_option(
            CreateApplicationCommandOption::default()
                .kind(CommandOptionType::SubCommand)
                .name("delete-all")
                .description("Deleta todos os seus tickets de um membro")
                .description_localized("en-US", "Delete all the tickets of a member")
                .add_sub_option(
                    CreateApplicationCommandOption::default()
                        .kind(CommandOptionType::User)
                        .name("member")
                        .description("O membro")
                        .description_localized("en-US", "The member")
                        .required(true)
                        .to_owned(),
                )
                .to_owned(),
        )
        .to_owned()
}

File: internal/commands/ticket/ticketadmin.go
package ticketcmds

import (
	"fmt"
	"log/slog"

	"github.com/bwmarrin/discordgo"
	"github.com/zanz1n/duvua-bot/internal/errors"
	"github.com/zanz1n/duvua-bot/internal/manager"
	"github.com/zanz1n/duvua-bot/internal/ticket"
	"github.com/zanz1n/duvua-bot/internal/utils"
)

var ticketadminCommandData = discordgo.ApplicationCommand{
	Name:        "ticketadmin",
	Type:        discordgo.ChatApplicationCommand,
	Description: "Comandos administrativos de tickets",
	DescriptionLocalizations: &map[discordgo.Locale]string{
		discordgo.EnglishUS: "Ticket administrative commands",
	},
	Options: []*discordgo.ApplicationCommandOption{
		{
			Type:        discordgo.ApplicationCommandOptionSubCommand,
			Name:        "enable",
			Description: "Habilita a funcionalidade de tickets no servidor",
			DescriptionLocalizations: map[discordgo.Locale]string{
				discordgo.EnglishUS: "Enables the ticket functionality on the server",
			},
		},
		{
			Type:        discordgo.ApplicationCommandOptionSubCommand,
			Name:        "disable",
			Description: "Desabilita a funcionalidade de tickets no servidor",
			DescriptionLocalizations: map[discordgo.Locale]string{
				discordgo.EnglishUS: "Disables the ticket functionality on the server",
			},
		},
		{
			Type:        discordgo.ApplicationCommandOptionSubCommand,
			Name:        "allow-multiple",
			Description: "Permite com que os membros criem mais de um ticket no servidor",
			DescriptionLocalizations: map[discordgo.Locale]string{
				discordgo.EnglishUS: "Allows members to create more than one ticket on the server",
			},
			Options: []*discordgo.ApplicationCommandOption{{
				Type:        discordgo.ApplicationCommandOptionBoolean,
				Name:        "allow",
				Description: "Permitir?",
				DescriptionLocalizations: map[discordgo.Locale]string{
					discordgo.EnglishUS: "Allow it?",
				},
				Required: true,
			}},
		},
		{
			Type:        discordgo.ApplicationCommandOptionSubCommand,
			Name:        "set-category",
			Description: "Define a categoria na qual o canal dos tickets serão criados",
			DescriptionLocalizations: map[discordgo.Locale]string{
				discordgo.EnglishUS: "Defines the category in which the ticket channels will be created",
			},
			Options: []*discordgo.ApplicationCommandOption{{
				Type:        discordgo.ApplicationCommandOptionChannel,
				Name:        "channel",
				Description: "A categoria de canais",
				DescriptionLocalizations: map[discordgo.Locale]string{
					discordgo.EnglishUS: "The channel category",
				},
			}},
		},
		{
			Type:        discordgo.ApplicationCommandOptionSubCommand,
			Name:        "add-permanent",
			Description: "Envia uma mensagem com um botão para criar tickets",
			DescriptionLocalizations: map[discordgo.Locale]string{
				discordgo.EnglishUS: "Posts a message with a button to create tickets",
			},
			Options: []*discordgo.ApplicationCommandOption{
				{
					Type:        discordgo.ApplicationCommandOptionChannel,
					Name:        "channel",
					Description: "O canal para enviar a mensagem",
					DescriptionLocalizations: map[discordgo.Locale]string{
						discordgo.EnglishUS: "The channel where the message will be sent",
					},
					Required: true,
				},
				{
					Type:        discordgo.ApplicationCommandOptionString,
					Name:        "message",
					Description: "A mensagem que será enviada",
					DescriptionLocalizations: map[discordgo.Locale]string{
						discordgo.EnglishUS: "The message that will be sent",
					},
				},
			},
		},
		{
			Type:        discordgo.ApplicationCommandOptionSubCommandGroup,
			Name:        "delete",
			Description: "Exclui tickets de membros",
			DescriptionLocalizations: map[discordgo.Locale]string{
				discordgo.EnglishUS: "Deletes members tickets",
			},
			Options: []*discordgo.ApplicationCommandOption{
				{
					Type:        discordgo.ApplicationCommandOptionSubCommand,
					Name:        "by-id",
					Description: "Exclui um ticket pelo seu id",
					DescriptionLocalizations: map[discordgo.Locale]string{
						discordgo.EnglishUS: "Deletes a ticket by it's id",
					},
					Options: []*discordgo.ApplicationCommandOption{{
						Type:        discordgo.ApplicationCommandOptionString,
						Name:        "id",
						Description: "O id do ticket que deseja excluir",
						DescriptionLocalizations: map[discordgo.Locale]string{
							discordgo.EnglishUS: "The id of the ticket you want to delete",
						},
						Required: true,
					}},
				},
				{
					Type:        discordgo.ApplicationCommandOptionSubCommand,
					Name:        "all",
					Description: "Exclui todos os tickets de um membro",
					DescriptionLocalizations: map[discordgo.Locale]string{
						discordgo.EnglishUS: "Deletes all the tickets of a member",
					},
					Options: []*discordgo.ApplicationCommandOption{{
						Type:        discordgo.ApplicationCommandOptionUser,
						Name:        "user",
						Description: "O membro cujos tickets serão excluídos",
						DescriptionLocalizations: map[discordgo.Locale]string{
							discordgo.EnglishUS: "The member whose tickets will be deleted",
						},
						Required: true,
					}},
				},
			},
		},
	},
}

func NewTicketAdminCommand(
	r ticket.TicketRepository,
	configR ticket.TicketConfigRepository,
) manager.Command {
	return manager.Command{
		Accepts: manager.CommandAccept{
			Slash:  true,
			Button: false,
		},
		Data:     &ticketadminCommandData,
		Category: manager.CommandCategoryConfig,
		Handler:  &TicketAdminCommand{r: r, configR: configR},
	}
}

type TicketAdminCommand struct {
	r       ticket.TicketRepository
	configR ticket.TicketConfigRepository
}

func (c *TicketAdminCommand) Handle(s *discordgo.Session, i *manager.InteractionCreate) error {
	if i.Member == nil || i.GuildID == "" {
		return errors.New("esse comando só pode ser utilizado dentro de um servidor")
	}
	if !utils.HasPerm(i.Member.Permissions, discordgo.PermissionAdministrator) {
		return errors.New("você não tem permissão para usar esse comando")
	}

	subCommandGroup := i.GetSubCommandGroup()
	subCommand, err := i.GetSubCommand()
	if err != nil {
		return err
	}

	var msg string
	if subCommandGroup != nil {
		switch subCommandGroup.Name {
		case "delete":
			switch subCommand.Name {
			case "by-id":
				id, e := i.GetStringOption("id", true)
				if e != nil {
					return e
				}
				return c.handleDeleteById(s, i, id)

			case "all":
				userId, e := i.GetUserOption("user", true)
				if e != nil {
					return e
				}

				if e = i.DeferReply(s, false); e != nil {
					return e
				}
				return c.handleDeleteAll(s, i, userId)

			default:
				return errors.New("opção `sub-command` inválida")
			}
		default:
			return errors.New("opção `sub-command-group` inválida")
		}
	} else {
		switch subCommand.Name {
		case "enable":
			msg, err = c.handleEnable(i.GuildID, true)

		case "disable":
			msg, err = c.handleEnable(i.GuildID, false)

		case "allow-multiple":
			allow, e := i.GetBooleanOption("allow", true)
			if e != nil {
				return e
			}
			msg, err = c.handleAllowMultiple(i.GuildID, allow)

		case "set-category":
			channelId, e := i.GetChannelOption("channel", false)
			if e != nil {
				return e
			}

			if channelId != "" {
				e = checkChannel(s, channelId, discordgo.ChannelTypeGuildCategory)
				if e != nil {
					return e
				}
			}
			msg, err = c.handleSetCategory(i.GuildID, channelId)

		case "add-permanent":
			channelId, e := i.GetChannelOption("channel", true)
			if e != nil {
				return e
			}
			e = checkChannel(s, channelId, discordgo.ChannelTypeGuildText)
			if e != nil {
				return e
			}

			message, e := i.GetStringOption("message", false)
			if e != nil {
				return e
			}
			if message == "" {
				message = "Clique no botão para criar o ticket"
			} else {
				message += fmt.Sprintf("\n- Mensagem por <@%s>", i.Member.User.ID)
			}

			msg, err = c.handleAddPermanent(s, channelId, message)

		default:
			return errors.New("opção `sub-command` inválida")
		}
	}

	if err != nil {
		return err
	}

	return i.Replyf(s, msg)
}

func (c *TicketAdminCommand) handleDeleteById(
	s *discordgo.Session,
	i *manager.InteractionCreate,
	id string,
) error {
	t, err := c.r.DeleteBySlug(id)
	if err != nil {
		return err
	}

	if t == nil {
		return errors.Newf("nenhum ticket `%s` encontrado", id)
	}

	comp := ""
	if _, err = s.ChannelDelete(t.ChannelId); err != nil {
		slog.Warn("Failed to delete ticket channel", "slug", t.Slug)
		comp += fmt.Sprintf(
			", mas não foi possível excluir o canal de texto <#%s>",
			t.ChannelId,
		)
	}

	// TODO: Dump ticket logs

	if t.ChannelId == i.ChannelID {
		return nil
	}

	return i.Replyf(s,
		"Ticket `%s` de <@%s> excluído com sucesso"+comp,
		t.Slug, t.UserId,
	)
}

func (c *TicketAdminCommand) handleDeleteAll(
	s *discordgo.Session,
	i *manager.InteractionCreate,
	userId string,
) error {
	ts, err := c.r.DeleteByMember(i.GuildID, userId)
	if err != nil {
		return err
	}

	if len(ts) == 0 {
		return errors.Newf("nenhum ticket de <@%s> encontrado", userId)
	}

	errc := []string{}
	containsChannel := false
	for _, t := range ts {
		if t.ChannelId == i.ChannelID {
			containsChannel = true
		}
		if _, err = s.ChannelDelete(t.ChannelId); err != nil {
			slog.Warn(
				"Failed to delete ticket channel",
				"guild_id", i.GuildID,
				"channel_id", t.ChannelId,
			)
			errc = append(errc, t.Slug)
		}
	}

	// TODO: Dump ticket logs

	if containsChannel {
		return nil
	}

	msg := fmt.Sprintf("%d tickets de <@%s> excluídos com sucesso", len(ts), userId)
	if len(errc) > 0 {
		msg += fmt.Sprintf(
			", mas %d canais de texto não puderam ser excluídos\n"+
				"São dos tickets: ",
			len(errc),
		)

		for i, s := range errc {
			if i != 0 {
				msg += ", "
			}
			msg += s
		}
	}

	return i.Replyf(s, msg)
}

func (c *TicketAdminCommand) handleEnable(guildId string, enabled bool) (string, error) {
	const (
		msgEnabled         = "**Tickets habilitados!**"
		msgAlreadyEnabled  = "**Os tickets já estavam habilitados!**"
		msgDisabled        = "**Tickets desabilitados!**"
		msgAlreadyDisabled = "**Os tickets já estavam desabilitados!**"
	)

	cfg, err := c.configR.GetByGuildId(guildId)
	if err != nil {
		return "", err
	}

	if cfg == nil {
		if enabled == ticket.DefaultConfigEnabled {
			return msgAlreadyDisabled, nil
		}
		_, err = c.configR.Create(ticket.TicketConfigCreateData{
			GuildId: guildId,
			Enabled: &enabled,
		})
		if err != nil {
			return "", err
		}
	} else {
		if cfg.Enabled == enabled {
			if enabled {
				return msgAlreadyEnabled, nil
			} else {
				return msgAlreadyDisabled, nil
			}
		}
		if err = c.configR.UpdateEnabled(guildId, enabled); err != nil {
			return "", err
		}
	}

	if enabled {
		return msgEnabled, nil
	} else {
		return msgDisabled, nil
	}
}

func (c *TicketAdminCommand) handleAllowMultiple(guildId string, allow bool) (string, error) {
	const (
		msgAllowed           = "**Agora um membro pode criar mais de um ticket**"
		msgAlreadyAllowed    = "**Membros já podiam criar mais de um ticket**"
		msgDisallowed        = "**Agora um mebro pode criar apenas um ticket por vez**"
		msgAlreadyDisallowed = "**Membros já podiam criar apenas um ticket por vez**"
		msgCompDisabled      = ", mas os tickets estão desabilitados no servidor.\n" +
			"Use `/ticketadmin enable` para habilitá-los."
	)

	cfg, err := c.configR.GetByGuildId(guildId)
	if err != nil {
		return "", err
	}

	if cfg == nil {
		if allow == ticket.DefaultConfigAllowMultiple {
			return msgAlreadyAllowed + msgCompDisabled, nil
		}

		_, err = c.configR.Create(ticket.TicketConfigCreateData{
			GuildId:       guildId,
			AllowMultiple: &allow,
		})
		if err != nil {
			return "", err
		}

		return msgDisallowed + msgCompDisabled, nil
	}

	if cfg.AllowMultiple == allow {
		var msg string
		if allow {
			msg = msgAlreadyAllowed
		} else {
			msg = msgAlreadyDisallowed
		}
		if !cfg.Enabled {
			msg += msgCompDisabled
		}
		return msg, nil
	}

	if err = c.configR.UpdateAllowMultiple(guildId, allow); err != nil {
		return "", err
	}

	var msg string
	if allow {
		msg = msgAllowed
	} else {
		msg = msgDisallowed
	}
	if !cfg.Enabled {
		msg += msgCompDisabled
	}

	return msg, nil
}

func (c *TicketAdminCommand) handleSetCategory(guildId, channelId string) (string, error) {
	const msgCompDisabled = ", mas os tickets estão desabilitados no servidor.\n" +
		"Use `/ticketadmin enable` para habilitá-los."

	cfg, err := c.configR.GetByGuildId(guildId)
	if err != nil {
		return "", err
	}

	if cfg == nil {
		cfg, err = c.configR.Create(ticket.TicketConfigCreateData{
			GuildId:           guildId,
			ChannelCategoryId: channelId,
		})
		if err != nil {
			return "", err
		}
	} else {
		if err = c.configR.UpdateChannelCategory(guildId, channelId); err != nil {
			return "", err
		}
	}

	var msg string
	if channelId == "" {
		msg = "**Os tickets serão criados sem uma categoria de canais**"
	} else {
		msg = fmt.Sprintf(
			"**Os tickets serão criados na categoria de canais <#%s>**",
			channelId,
		)
	}

	if !cfg.Enabled {
		msg += msgCompDisabled
	}

	return msg, nil
}

func (c *TicketAdminCommand) handleAddPermanent(
	s *discordgo.Session,
	channelId, message string,
) (string, error) {
	msg := discordgo.MessageSend{
		Content: message,
		Components: []discordgo.MessageComponent{discordgo.ActionsRow{
			Components: []discordgo.MessageComponent{discordgo.Button{
				Label:    "Criar ticket",
				Style:    discordgo.PrimaryButton,
				Emoji:    emoji("🎫"),
				CustomID: "ticket/create",
			}},
		}},
	}

	_, err := s.ChannelMessageSendComplex(channelId, &msg)
	if err != nil {
		return "", err
	}

	return "Mensagem enviada com sucesso", nil
}

func checkChannel(s *discordgo.Session, id string, kind discordgo.ChannelType) error {
	ch, e := s.State.Channel(id)
	if e != nil {
		ch, e = s.Channel(id)
	}
	if e != nil || ch.Type != kind {
		switch kind {
		case discordgo.ChannelTypeGuildText:
			return errors.New("opção `channel` precisa ser um canal de texto válido")
		case discordgo.ChannelTypeGuildCategory:
			return errors.New("opção `channel` precisa ser uma categoria de canais válida")
		default:
			return errors.New("opção `channel` precisa ser um canal válido")
		}
	}
	return nil
}

func emoji(name string) *discordgo.ComponentEmoji {
	return &discordgo.ComponentEmoji{Name: name}
}
